plugin/pkg/admission/resourcedefaults/admission.go

```diff
@@ -0,0 +1,70 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resourcedefaults
+
+import (
+	"io"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/admission"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/resource"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+)
+
+func init() {
+	admission.RegisterPlugin("ResourceDefaults", func(client client.Interface, config io.Reader) (admission.Interface, error) {
+		return NewResourceDefaults(), nil
+	})
+}
+
+const (
+	defaultMemory string = "512Mi"
+	defaultCPU    string = "1"
+)
+
+// resourceDefaults is an implementation of admission.Interface which applies default resource limits (cpu/memory)
+// It is useful for clusters that do not want to support unlimited usage constraints, but instead supply sensible defaults
+type resourceDefaults struct{}
+
+func (resourceDefaults) Admit(a admission.Attributes) (err error) {
+	// ignore deletes, only process create and update
+	if a.GetOperation() == "DELETE" {
+		return nil
+	}
+
+	// we only care about pods
+	if a.GetKind() != "pods" {
+		return nil
+	}
+
+	// get the pod, so we can validate each of the containers within have default mem / cpu constraints
```

Comment:
I am out of office today and tomorrow, but will try to take a look later
today.
On Jan 15, 2015 8:27 AM, "Eric Tune" <notifications@github.com> wrote:

> In plugin/pkg/admission/resourcedefaults/admission.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3472#discussion-diff-23021806>
> :
>
> > +// resourceDefaults is an implementation of admission.Interface which applies default resource limits (cpu/memory)
> > +// It is useful for clusters that do not want to support unlimited usage constraints, but instead supply sensible defaults
> > +type resourceDefaults struct{}
> > +
> > +func (resourceDefaults) Admit(a admission.Attributes) (err error) {
> > +	// ignore deletes, only process create and update
> > +	if a.GetOperation() == "DELETE" {
> > +		return nil
> > +	}
> > +
> > +	// we only care about pods
> > +	if a.GetKind() != "pods" {
> > +		return nil
> > +	}
> > +
> > +	// get the pod, so we can validate each of the containers within have default mem / cpu constraints
>
> This seems closely related to (#2587
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/2587>). Setting
> defaults is a tricky business. I think we should let @thockin
> <https://github.com/thockin> comment. He might say that we need to reach
> agreement on how to default values first before adding code that defaults
> values, or that this is simple enough to not need to wait. He might say
> that this is really a type of validation or conversion function, and
> belongs there, or not.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3472/files#r23021806>
> .
>